pkg/splunk/enterprise/clustermaster.go

```diff
@@ -83,33 +83,44 @@ func ApplyClusterMaster(client splcommon.ControllerClient, cr *enterprisev1.Clus
 		}
 	}()
 
+	// Register the S3 clients specific to providers if not done already
+	// This is done to prevent the null pointer dereference in case when
+	// operator crashes and comes back up and the status of app context was updated
+	// to match the spec in the previous run.
+	if cr.Spec.AppFrameworkConfig.VolList != nil {
+		RegisterS3ClientsForProviders(cr.Spec.AppFrameworkConfig.VolList)
+	}
+
 	//check if the apps need to be downloaded from remote storage
-	if cr.Spec.CommonSplunkSpec.Mock != true && !reflect.DeepEqual(cr.Status.AppContext.AppFrameworkConfig, cr.Spec.AppFrameworkConfig) {
+	if ShouldCheckAppStatus(cr.Spec.AppFrameworkConfig, cr.Status.AppContext.AppInfoStatus) || !reflect.DeepEqual(cr.Status.AppContext.AppFrameworkConfig, cr.Spec.AppFrameworkConfig) {
 		var sourceToAppsList map[string]splclient.S3Response
 
-		for _, vol := range cr.Spec.AppFrameworkConfig.VolList {
-			if _, ok := splclient.S3Clients[vol.Provider]; !ok {
-				splclient.RegisterS3Client(vol.Provider)
-			}
-		}
+		scopedLog.Info("Checking status of apps on remote storage...")
 
+		shouldHandleAppChanges := true
 		sourceToAppsList, err = GetAppListFromS3Bucket(client, cr, &cr.Spec.AppFrameworkConfig)
 		if len(sourceToAppsList) != len(cr.Spec.AppFrameworkConfig.AppSources) {
 			scopedLog.Error(err, "Unable to get apps list, will retry in next reconcile...")
-			return result, err
+			shouldHandleAppChanges = false
 		}
 
 		for _, appSource := range cr.Spec.AppFrameworkConfig.AppSources {
 			scopedLog.Info("Apps List retrieved from remote storage", "App Source", appSource.Name, "Content", sourceToAppsList[appSource.Name].Objects)
 		}
 
-		err = handleAppRepoChanges(client, cr, &cr.Status.AppContext, sourceToAppsList, &cr.Spec.AppFrameworkConfig)
-		if err != nil {
-			scopedLog.Error(err, "Unable to use the App list retrieved from the remote storage")
-			return result, err
+		// Only handle the app repo changes if we were able to successfully get the apps list
+		if shouldHandleAppChanges == true {
+			err = handleAppRepoChanges(client, cr, &cr.Status.AppContext, sourceToAppsList, &cr.Spec.AppFrameworkConfig)
+			if err != nil {
+				scopedLog.Error(err, "Unable to use the App list retrieved from the remote storage")
+				return result, err
+			}
+
+			cr.Status.AppContext.AppFrameworkConfig = cr.Spec.AppFrameworkConfig
```

Comment:
We can move it to the above scope block when the expected and received count matches for the map entries.??

Reply:
Sorry did not understand this. The above if condition is for when the expected and received count does not match.